### DISCUSSION TITLE
C++20 support

### DISCUSSION THREAD

**Discussion by lbt0116** [Simple-Robotics/aligator#254] [General]
The project currently faces compatibility issues with C++20, such as disallowing template parameters in destructors (e.g., ~ConstraintSetTpl<Scalar>()) and the use of non-constexpr types like std::string in constant expressions, which prevent leveraging modern C++20 features. To address this, I propose ensuring full C++20 compliance by updating problematic code, replacing non-compliant patterns, and adding explicit C++20 support in the build system (e.g., CMake flags). This would enable developers to benefit from C++20's improvements like concepts, ranges, and enhanced constexpr, ensuring the library stays modern and developer-friendly.

**Comment by ManifoldFR** [comment]
Hello, thanks for your interest in aligator!

Could you elaborate on some of these points? I am not familiar with C++20 beyond concepts.  
We haven't looked at implementing C++20 features yet, due to both lack of time and lack of return on investment: inherent problems in even several robotics software stacks (looking at you, ROS) limited us to C++14 when we first started. Getting to C++17 was already a difficult negotiation.

We could get onto it with your help and the help of other developers if we could get a report (and a PR down the line) on e.g. noncompliant code and how to fix the CMake build flags.

**Comment by ManifoldFR** [comment]
Hello @lbt0116, would you have time to come back to this? I will be converting this to a discussion in the meantime.

**Reply by lbt0116** [reply]
I apologize for the delayed response. Initially, I intended to integrate Aligator into my code environment, utilizing C++20's atomic shared pointers for data sharing. However, I discovered that Aligator has conflicts with the C++20 standard. As a result, I later restructured the program into a distributed architecture using ROS 2, running Aligator in a separate node process to resolve the compatibility issues.

**Reply by ManifoldFR** [reply]
No worries!

I did not know about [this feature](https://en.cppreference.com/w/cpp/memory/shared_ptr/atomic2). This allows to avoid race conditions when accessing `shared_ptr` ?

What is your use case here, modifying aligator problems or accessing objects in the problem workspace (a lot of shared_ptr in there) from another thread?

Do you remember which issues you had when trying to compile aligator in C++20 ? In both CMake and the C++ code itself? I might have some projects using aligator together C++20 coming down the line so it might also be useful to some of us here at simple-robotics.

**Reply by lbt0116** [reply]
Atomic smart pointers can provide lock-free operations. While shared_ptr ensures thread safety only for reference counting, atomic smart pointers guarantee complete thread safety. My usage involved storing Aligator's inputs and outputs in atomic smart pointers to facilitate data sharing in a multi-threaded environment. When compiling Aligator with C++20, I recall encountering an issue like "template-id not allowed for destructor." This behavior was modified in C++20. Therefore, isolating Aligator and other algorithms using C++20 into separate processes seems to be a better choice. ROS 2 provides an intra-process communication option, which facilitates efficient communication between different processes on the same host.

**Reply by ManifoldFR** [reply]
A quick regex search found where that could come from, upstream inside [proxsuite-nlp/constraint-set.hpp](https://github.com/Simple-Robotics/proxsuite-nlp/blob/f9f45ebc37257881bf05bcd04d648b8dd90edded/include/proxsuite-nlp/constraint-set.hpp#L80). Here the destructor of `ConstraintSetTpl` is qualified using a template parameter, which should be wrong and is the only instance of this across aligator & proxsuite-nlp. I'll push a commit upstream. Thanks!

**Reply by ManifoldFR** [reply]
Can you try compiling proxsuite-nlp (commit 6c2ed54c7bb8d4e7f72df4e5e38a57a16e2793a8) + aligator with your setup?

**Comment by lbt0116** [comment]
I compiled the version you provided, and indeed the previous ConstraintSetTpl issue has disappeared. However, there is still a problem related to fmt, as shown below:

kinodynamics-fwd.hxx:27:5: error：call to non-‘constexpr’ function ‘std::string fmt::v11::format(format_string<T ...>, T&& ...) [with T = {long unsigned int, long unsigned int}; std::string = std::__cxx11::basic_string<char>; format_string<T ...> = fstring<long unsigned int, long unsigned int>]’
   27 |     ALIGATOR_DOMAIN_ERROR(
      |     ^~~~~~~~~~~~~~~~~~~~~

**Reply by lbt0116** [reply]
I suspect that the std::fmt introduced in C++20 is conflicting with the currently used fmt library.

**Reply by ManifoldFR** [reply]
This is related to https://github.com/fmtlib/fmt/issues/2455 it seems. There is a conflict with C++20's consteval.

https://github.com/chemfiles/chemfiles/issues/492 seems to provide a workaround?

Can you check if #255 might fix your issue?

**Reply by lbt0116** [reply]
I will try it tomorrow and give you feedback😄

**Reply by lbt0116** [reply]
I can not find refs/heads/topic/domain-errors-va-args with exit status 128.

**Reply by ManifoldFR** [reply]
It was merged into devel yesterday, so you can just try devel :)

**Reply by lbt0116** [reply]
Thx for your help, it can run at c++20 right now😄

**Reply by ManifoldFR** [reply]
Great ! Please tell us if you have another issue later on, aligator is still under active development :)
